[PLAT-665] Add feedback button to Quick Files

sloria opened the pull request:
<!-- Before you submit your Pull Request, make sure you picked the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features and non-hotfix bugfixes, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

Add a feedback button to the Quick Files page.

<img width="500" alt="steven loria s quick files 2018-02-05 23-20-18" src="https://user-images.githubusercontent.com/2379650/35841669-2f305ce8-0acb-11e8-8ed9-fd2030c41bdf.png">

<img width="500" alt="steven loria s quick files 2018-02-06 08-29-31" src="https://user-images.githubusercontent.com/2379650/35861941-eefaed8a-0b17-11e8-80d3-6194bd6273bc.png">



## Summary of Changes

- Add microfeedback-button as a dependency
- Add `feedback-button` component

## Side Effects / Testing Notes

- Requires `MICROFEEDBACK_URL` envvar to be set. Use `'https://microfeedback-osf.herokuapp.com/13209/3'` for staging and prod. @icereval 

## Ticket

https://openscience.atlassian.net/browse/PLAT-665

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] ~~changes described in `CHANGELOG.md`~~ *(don't have initial release yet)*

<!-- Please strike through any checks that you think are not relevant for this PR and indicate why, e.g.

     - [ ] ~~easy to understand~~ *(necessarily complex)*
-->

aaxelb commented on app/styles/_global.scss:
```diff
@@ -3,3 +3,14 @@
     opacity: 1;
 }
+
+// microfeedback-button styling
```
The auto-generated class is available as `styleNamespace` on the component -- if you apply that to the alert you could move these styles to `components/feedback-button/styles.scss`

sloria replied:
@aaxelb The markup is generated by microfeedback-button, not by ember, so we can't use the auto-generated classes.

sloria replied:
That said, I suppose I could add an option to `microfeedback-button` so that you could add a custom class. Something like

```js
microfeedback({
  className: 'OSFMicroFeedbackButton',
})
```

```scss
& .OSFMicroFeedbackButton {
   ...
}
```

But that can be a future improvement.

I think the real fix would be to set the correct base font size on the `html` element. For some reason, ember-bootstrap (I think?) seems to be setting `font-size` to `10px` on `html` then overrides it to `14px` on `body`. Since microfeedback uses `rem` for font sizes, the text ends up being too small, hence these overrides.

If we could instead set `font-size` to `100%` on `html`, the CSS in this PR would be unnecessary. That might have unintended consequences, so I decided not to do that in this PR.

aaxelb replied:
Heh, I was already submitting a [feature request](https://github.com/microfeedback/microfeedback-button/issues/26). You would still have to use `styleNamespace`, though:
```js
microfeedback({
  className: this.get('styleNamespace'), // or maybe `${this.get('styleNamespace')}__microfeedback`
})
```

We should definitely set font size correctly, and probably use `rem` throughout. But that might fall under the [responsive font size analysis](https://openscience.atlassian.net/browse/EMB-91). I'm fine with these global styles for now.

jamescdavis commented on app/components/feedback-button/component.js:
```diff
@@ -0,0 +1,68 @@
+            if (this.get('priorityID')) {
+                params.priorityID = this.get('priorityID');
+            }
+            const query = $.param(params);
```
Make sure you `import $ from 'jquery';`

sloria commented on app/user-quickfiles/template.hbs:
```diff
@@ -9,3 +9,5 @@
     {{file-browser user=model openFile=(action 'openFile') dropZoneId='quickfiles-dropzone'}}
 </div>
+
+{{feedback-button user=model text='Tell us what you think of the new Quick Files page.' componentID='13836' priorityID='10100'}}
```
> This should not use a bare string. This text should be moved to the translations file.

Will do.

> componentID and priorityID should not be hardcoded here. These should come from the environment.

These are specific to this page, not the environment, which is why I put them here.

sloria commented on app/styles/_global.scss:
```diff
@@ -3,3 +3,14 @@
+
+.microfeedback-button { // stylelint-disable-line selector-class-pattern
+    font-size: 1em !important;
+}
```
See https://github.com/CenterForOpenScience/ember-osf-web/pull/67#discussion_r166322314 .

sloria commented on app/components/feedback-button/component.js:
```diff
@@ -0,0 +1,68 @@
+                    <label style="font-weight: 300;" id="microfeedback-followup" for="swal2-checkbox" class="swal2-checkbox">
+                    <input style="display:block" type="checkbox">
+                    <span>${followUpLabel}</span>
+                    </label>`,
```
> This large block of HTML is hard to read. HTML should go in templates.

Does ember-cli have a way of loading HTML files?

> The inline styles should be moved to a stylesheet and proper classes added to elements.

OK, I can move those.

sloria replied:
> Injecting variables into an HTML string is risky. Template variables are sanitized so bad things don't happen when they contain HTML tags.

No user input is being used here, but I understand the concern. Do we have code for sanitizing? We use this in OSF to escape HTML entities:

```js
/**
 * Escapes html characters in a string.
 */
var htmlEscape = function(text) {
    return $('<div/>').text(text).html();
};

htmlEscape('<script>alert("foo")</script>')
//  "&lt;script&gt;alert(\"foo\")&lt;/script&gt;"
```

sloria commented on app/components/feedback-button/component.js:
```diff
@@ -0,0 +1,68 @@
+            backgroundColor: buttonColor,
+            showDialog: (btn) => {
+                const text = this.get('text') || '';
+                const placeholder = this.get('placeholder') || 'Describe your issue or share your ideas';
```
I can make it so that the placeholder attribute isn't added if the `placeholder` component prop is falsy. I didn't initially add this because it complicated the code and didn't think it'd be a use case for us (all the placeholders for these feedback dialogs should be consistent anyway).

I can certainly implement that, though.

sloria commented:
> If we are going to use microfeedback in Ember, we should create an Ember front-end for it. At the very least, it should be wrapped with an Ember addon that is maintained separately. What I would like to see is a block component where you wrap the HTML in the template, e.g. something like:

Yeah, I agree this implementation isn't ideal, and a proper Ember component would be better. I was shooting for 80:20 here, in hopes that we could start capturing user feedback immediately, with as little added code as possible. I understand that we shouldn't stray far from Ember patterns, though.

> The biggest issues I have with it is the big block of html outside of a template

Yeah, this is a limitation of the underlying library, sweetalert2, which requires passing the HTML in order to do multiple inputs: https://sweetalert2.github.io/#multiple-inputs .
In discussing this with product, we really wanted the "Contact me to follow up" checkbox, so we needed the multiple inputs.

> and using a different kind of modal than we use everywhere else. 

Also a fair point. Again, I was going for as minimal code as possible; I didn't spend a lot of effort on styling past sweetalert2's defaults. I can work on this more, though. This example from the sweetalert2 docs uses bootstrap button classes: https://sweetalert2.github.io/#dismiss-handle . Not sure if that's worth it, though. As you pointed out, the better long-term solution is a proper Ember component.

That said, it's relatively common to have these auxiliary feedback/customer support UI components (e.g., Intercom, Drift, UserVoice) to be branded/styled differently from the app they exist in. Not that we should take bad practices from other companies =) .

@jamescdavis I'll take a stab at making the widget an Ember component if I have some extra time today.

sloria commented:
I've addressed the nits.

Also, the buttons are now Bootstrap-styled.

<img width="400" alt="steven loria s quick files 2018-02-06 15-22-37" src="https://user-images.githubusercontent.com/2379650/35882308-b53cc37c-0b51-11e8-9ac1-e5aff884aeb6.png">

I probably won't have time to convert this into an Ember component today, but I'll work on that in a separate branch. 

@jamescdavis  It's your call if you want to merge this as is.

jamescdavis commented on app/components/feedback-button/component.js:
```diff
@@ -0,0 +1,90 @@
+                        <label id="microfeedback-followup" for="swal2-checkbox" class="${styleNamespace}CheckboxLabel swal2-checkbox">
+                            <input class="${styleNamespace}Checkbox" type="checkbox">
+                            <span>${followUpLabel}</span>
+                        </label>
```
See: #71

sloria commented on app/user-quickfiles/template.hbs:
```diff
@@ -9,3 +9,5 @@
     {{file-browser user=model openFile=(action 'openFile') dropZoneId='quickfiles-dropzone'}}
 </div>
+
+{{feedback-button user=model text=(t 'quickfiles.feedback_dialog_text') componentID='13836' priorityID='10100'}}
```
Good idea. I've done most of the work to make the widget an ember component. I'll send a new PR and address this comment there.

sloria commented:
OK, this is replaced by #74
